### PR TITLE
Fix amqp reconnection

### PR DESCRIPTION
**Pull request by zhouzhuojie**
Any network or server-side connection failures will stop the amqp's worker. This pr adds reconnection to fix the issue.

### CODE REVIEW

File: amqp.go [hunk 1]
```diff
@@ -47,46 +42,51 @@ func publishToAMQP(amqpURL string, exchange string, routingKey string, payload s
...
 	for amqp, ms := range mocks {
-		func(amqp ExpectAMQP, ms MocksArray) {
-			ch := getAMQPChannel(om.AMQPURL)
-			if err := prepareChannel(ch, amqp.Exchange, amqp.RoutingKey, amqp.Queue); err != nil {
-				logrus.Errorf("%s: %s", "failed to prepare the channel for amqp", err)
-				return
-			}
+		go func(amqp ExpectAMQP, ms MocksArray) {
+			// use a for loop here to reconnect on failures
+			for {
```
Comment (ziru) [inline amqp.go]: can we add a INFO log statement here?

File: amqp.go [hunk 2]
```diff
@@ -26,7 +19,9 @@ func publishToAMQP(amqpURL string, exchange string, routingKey string, payload s
 		}).Info("try to publish to amqp")
 	}()
 
-	ch := getAMQPChannel(amqpURL)
+	ch, cleanUp := newAMQPChannel(amqpURL)
+	defer cleanUp()
+
```
Comment (ziru) [inline amqp.go]: rather than dialing per every publish, can we retry on publish failures (and re-dial only on publish failures)?

Comment (zhouzhuojie) [reply amqp.go]: we can, but it requires some mutex locking for connection recreation and swapping, not sure if it worth the effort for low volume traffic mocking.

Also, having a new connection is not that expensive.

File: amqp.go [hunk 3]
```diff
@@ -47,59 +42,68 @@ func publishToAMQP(amqpURL string, exchange string, routingKey string, payload s
...
+				"exchange":    msg.Exchange,
+				"routing_key": msg.RoutingKey,
+			}).Info()
+			ms.DoActions(&Context{
+				AMQPPayload: string(msg.Body),
+				om:          om,
+			})
+		}
+
+		logrus.WithFields(logrus.Fields{"queue": amqp.Queue}).Warn("amqp worker connection reset")
```
Comment (zhouzhuojie) [inline amqp.go]: @ziru added a warning about the connection being reset

File: amqp.go (hunk 2 above)
Comment (ziru) [reply amqp.go]: fair enough

**Pull request merged**